zcash_primitives/src/transaction/components/sapling.rs

```diff
@@ -83,13 +83,52 @@ impl MapAuth<Authorized, Authorized> for () {
 
 #[derive(Debug, Clone)]
 pub struct Bundle<A: Authorization> {
-    pub shielded_spends: Vec<SpendDescription<A>>,
-    pub shielded_outputs: Vec<OutputDescription<A::Proof>>,
-    pub value_balance: Amount,
-    pub authorization: A,
+    shielded_spends: Vec<SpendDescription<A>>,
+    shielded_outputs: Vec<OutputDescription<A::Proof>>,
+    value_balance: Amount,
+    authorization: A,
 }
 
 impl<A: Authorization> Bundle<A> {
+    /// Constructs a `Bundle` from its constituent parts.
+    pub(crate) fn from_parts(
+        shielded_spends: Vec<SpendDescription<A>>,
+        shielded_outputs: Vec<OutputDescription<A::Proof>>,
+        value_balance: Amount,
+        authorization: A,
+    ) -> Self {
+        Bundle {
+            shielded_spends,
+            shielded_outputs,
+            value_balance,
+            authorization,
+        }
+    }
+
+    /// Returns the list of spends in this bundle.
+    pub fn shielded_spends(&self) -> &[SpendDescription<A>] {
+        &self.shielded_spends
+    }
+
+    /// Returns the list of outputs in this bundle.
+    pub fn shielded_outputs(&self) -> &[OutputDescription<A::Proof>] {
```

Comment:
Zingolib's test framework constructs mock bundles, and adds outputs to mock bundles, could we get a getter fn for a `mut &` to the vec of outputs (probably behind a `#[cfg(feature = "test-dependences")]`), and public access to `from_parts`? Probably something like the `OutputDescription` test impl?

Reply:
Perhaps instead we should have an API for constructing these mock bundles. We do already have `MockTxProver` which means you can use `SaplingBuilder` to create fake Sapling bundles with invalid proofs while everything else is valid. But that doesn't let you construct arbitrarily invalid bundles (as this would necessarily violate type safety).

Can you link me to how you currently do this?

Reply:
We create one here: https://github.com/zingolabs/zingolib/blob/8bcad1020357047b1fbe4915e102868a45aa0280/lib/src/lightclient/testmocks.rs#L9-L14 and https://github.com/zingolabs/zingolib/blob/6ad127e73a41f7bfd612597677b292437812f7d1/lib/src/blaze/test_utils.rs#L119-L124

And we add outputs here: https://github.com/zingolabs/zingolib/blob/6ad127e73a41f7bfd612597677b292437812f7d1/lib/src/blaze/test_utils.rs#L217

Reply:
Mmk, all of these examples could definitely be replaced by usage of `SaplingBuilder` (see the changes to [zcash_primitives/benches/note_decryption.rs](https://github.com/zcash/librustzcash/pull/734/files#diff-09d5a6ce4867b5daf6d0c6b4fb7815ab80432d14c20025660ef9a625747b7253R40-R54) in this PR). However, once we pull this code out into `sapling-crypto`, we will need `Bundle::from_parts` to be public (the same way it is for `orchard`) in order for transaction parsing logic to remain in `zcash_primitives`. So I'd be okay with exposing that.

I am not keen on exposing `&mut` references to the internals of a `Bundle`; the only reason those methods exist on `OutputDescription` is for the purpose of testing the internals of note decryption, which is implemented as part of the crate.